### PR TITLE
Refactor/update wdm dependent plugins

### PR DESCRIPTION
**Pull request by InteractiveTimmy**
# Pull Request

## Description

The changes in this pull request are scoped to migrating all packages to utilize the new `internal-plugin-device` plugin. Note on these failing tests on both the `master` branch and this branch:

Plugins without test failures:
```
webex
@webex/http-core
@webex/webex-core
@webex/internal-plugin-avatar
@webex/internal-plugin-calendar
@webex/internal-plugin-flag
@webex/internal-plugin-mercury
@webex/internal-plugin-metrics
@webex/internal-plugin-presence
@webex/internal-plugin-support
@webex/internal-plugin-team
@webex/plugin-authorization-browser
@webex/plugin-authorization-node
@webex/plugin-team-memberships
@webex/plugin-meetings
@webex/plugin-teams
@webex/plugin-webhooks
```

Plugins without real tests:
```
@webex/plugin-device-manager
@webex/recipe-private-web-client
```

Plugins with test failures:
`@webex/internal-plugin-encryption` -1
These tests are failing on both master and this implementation branch. One less error on this implementation branch.
```
1) Encryption
       KMS
         when ecdhe negotiation times out
           when ecdhe is renegotiated
             limits the number of retries:
     KmsTimeoutError: The KMS did not respond within 400 milliseconds
KMS_REQUEST: update /ping
KMS_REQUEST_ID: 71d6f0c9-04be-4883-a5ef-cc194cb9bb4c
    at KmsTimeoutError.ExtendableBuiltin (packages/node_modules/@webex/common/src/exception.js:50:44)
    at KmsTimeoutError.Exception (packages/node_modules/@webex/common/src/exception.js:97:142)
    at KmsTimeoutError.KmsError (packages/node_modules/@webex/internal-plugin-encryption/src/kms-errors.js:54:123)
    at new KmsTimeoutError (packages/node_modules/@webex/internal-plugin-encryption/src/kms-errors.js:114:137)
    at Timeout.<anonymous> (packages/node_modules/@webex/internal-plugin-encryption/src/kms-batcher.js:77:40)
```

`@webex/internal-plugin-feature` 0
The test is marked as flakey, and it appears to exist on both the master branch and the dev branch.
```
1) plugin-feature
       when feature toggle update event is received
         updates the feature toggle:
     Error: event:featureToggle_update did not fire within 10000ms
    at Timeout._onTimeout (packages/node_modules/@webex/test-helper-mocha/src/index.js:337:16)
```

`@webex/internal-plugin-search` -2 ???
This package has two fewer failures on this branch in contrast to master.
```
FAILED TESTS:
  plugin-search
    #search
11 02 2020 12:24:16.801:WARN [reporter]: SourceMap position not found for trace: ExtendableBuiltin@http://localhost:8001/absolute/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/515d3e9d743e2e22edde0dcb1d97a488.browserify.js?cee46de35d1aad16a6b2e81006b695f67a5f9d54:233387:44
11 02 2020 12:24:16.802:WARN [reporter]: SourceMap position not found for trace: HttpError@http://localhost:8001/absolute/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/515d3e9d743e2e22edde0dcb1d97a488.browserify.js?cee46de35d1aad16a6b2e81006b695f67a5f9d54:236369:125
11 02 2020 12:24:16.802:WARN [reporter]: SourceMap position not found for trace: WebexHttpError@http://localhost:8001/absolute/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/515d3e9d743e2e22edde0dcb1d97a488.browserify.js?cee46de35d1aad16a6b2e81006b695f67a5f9d54:260881:135
      ✖ "before all" hook: create a room for kirkEU in "#search"
        HeadlessChrome 80.0.3987 (Mac OS X 10.15.3)
        Firefox 72.0.0 (Mac OS X 10.15.0)
      Request is only permitted in user's home teams cluster
      POST https://conversation-intb.ciscospark.com/conversation/api/v1/conversations
      WEBEX_TRACKING_ID: webex-js-sdk_ad5acd88-0a40-4083-b7cd-2c01a5af3f05_7
```

`@webex/internal-plugin-user` +0
These errors exist on master and seem to be related to how the service catalog isn't ready by the time the test runs.
```
1) plugin-user with US user without Federation enabled
       #activate()
         retrieves a valid user token and sets the password:
     Error: service-interceptor: 'atlas' is not a known service
    at /Users/timsch/Projects/webex-sdks/webex-js-sdk/packages/node_modules/@webex/webex-core/src/lib/services/interceptors/service.js:88:41
  

  2) plugin-user with US user with Federation enabled
       #activate()
         retrieves a valid user token and sets the password:
     Error: service-interceptor: 'atlas' is not a known service
    at /Users/timsch/Projects/webex-sdks/webex-js-sdk/packages/node_modules/@webex/webex-core/src/lib/services/interceptors/service.js:88:41
```

`@webex/plugin-phone` -17
This plugin appears flakey, but gives less failed tests now that the new federated `internal-plugin-device` is being used.
```
FAILED TESTS:
      Events Model
        a room identified by sip uri
          ✖ proceeds through a series of events
            Firefox 73.0.0 (Mac OS X 10.15.0)
            HeadlessChrome 80.0.3987 (Mac OS X 10.15.3)
          TypeError: Cannot read property 'length' of undefined
              at Object.decode (/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/node_modules/urlsafe-base64/lib/urlsafe-base64.js:53:1 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:242086:30)
              at fromBase64url (/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/common/dist/base64.js:24:32 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:246766:32)
              at Object.decode (/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/common/dist/base64.js:61:10 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:246803:10)
              at /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/plugin-phone/dist/call.js:1124:37 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:266605:37
              at new Promise (<anonymous>)
              at new F (/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/node_modules/core-js/library/modules/_export.js:36:1 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:24989:28)
              at child.getTargetFromInvitee (/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/plugin-phone/dist/call.js:1121:12 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:266602:12)
              at /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/plugin-phone/dist/call.js:1098:21 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:266580:21

          ✖ lists the existing call
            Firefox 73.0.0 (Mac OS X 10.15.0)
          expected 1 to equal 2
          AssertionError@/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/node_modules/assertion-error/index.js:74:1 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:7970:13
          require<[107]</module.exports/Assertion.prototype.assert@/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/node_modules/chai/lib/chai/assertion.js:141:1 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:13262:13
          require<[110]</module.exports/assert.equal@/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/node_modules/chai/lib/chai/interface/assert.js:139:1 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:17329:10
          require<["/Users/timsch/Projects/webex-sdks/webex-js-sdk/packages/node_modules/@webex/plugin-phone/test/integration/spec/phone.js"]</</</</</</<@/var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/packages/node_modules/@webex/plugin-phone/test/integration/spec/phone.js:346:22 <- /var/folders/s_/371p9cg118qdxr2lnj6hj8r00000gn/T/74e05a47e6510022e0251952bd361108.browserify.js:272956:36
```

`@webex/webex-server` +2
These errors are related to how the new devices plugin no longer has a `services` property. These tests need to be refactored to no longer check for this property.
```
1) /api/v1/session
       PUT
         authorizes a webex instance with the specified user:
     AssertionError: expected { Object (name, model, ...) } to have property 'services'
    at expectSession (packages/node_modules/@webex/webex-server/test/integration/spec/session.js:43:26)
    at Test._assertFunction (node_modules/supertest/lib/test.js:283:11)
    at Test.assert (node_modules/supertest/lib/test.js:173:18)
    at Server.localAssert (node_modules/supertest/lib/test.js:131:12)
    at Server.EventEmitter.emit (domain.js:448:20)
    at emitCloseNT (net.js:1619:8)
    at process._tickCallback (internal/process/next_tick.js:63:19)
  

  2) /api/v1/session/invoke*
       POST
         returns the result of the method at the specified keypath:
     AssertionError: expected { Object (name, model, ...) } to have property 'services'
    at expectSession (packages/node_modules/@webex/webex-server/test/integration/spec/session.js:43:26)
    at Test._assertFunction (node_modules/supertest/lib/test.js:283:11)
    at Test.assert (node_modules/supertest/lib/test.js:173:18)
    at Server.localAssert (node_modules/supertest/lib/test.js:131:12)
    at Server.EventEmitter.emit (domain.js:448:20)
    at emitCloseNT (net.js:1619:8)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

Fixes # [SPARK-119622](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-119622)

## Type of Change

- [x] Refactor.

## Test Coverage

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Ran all tests to validate that the changes did not impact any existing tests.

**Test Configuration**:
* Node/Browser Version v10.18.0
* NPM Version v6.13.4

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js [hunk 1]
```diff
@@ -67,135 +67,68 @@ describe('plugin-device', () => {
...
-            return services.updateServices();
-          });
+          services.updateServices();
+        });
 
+        describe('when the \'wdm\' service does exist after wait', () => {
           it('should return a resolved promise', () => {
-            assert.isFulfilled(device.canRegister(wait));
+            /* eslint-disable-next-line no-underscore-dangle */
+            assert.isTrue(services._getCatalog().status.postauth.collecting);
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: What is the purpose of this assert? It isn't described in the test block.

File: packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js [hunk 2]
```diff
@@ -67,135 +67,68 @@ describe('plugin-device', () => {
...
-            services.updateServices();
+        describe('when the \'wdm\' service does not exist after wait', () => {
+          beforeEach('remove wdm service', () => {
+            services.get = sinon.stub().returns(undefined);
+            services.waitForCatalog = sinon.stub().resolves();
           });
 
           it('should return a rejected promise', () => {
             /* eslint-disable-next-line no-underscore-dangle */
             assert.isTrue(services._getCatalog().status.postauth.collecting);
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: This assertion isn't described in the test block.

File: packages/node_modules/@webex/internal-plugin-device/src/device.js [hunk 3]
```diff
@@ -468,38 +468,25 @@ const Device = WebexPlugin.extend({
...
 
     // Destructure the services plugin for ease of reference.
     const {services} = this.webex.internal;
 
-    // Validate if the service url exists in the service catalog.
-    if (services.get('wdm')) {
-      return Promise.resolve();
-    }
-
     // If wait is true, wait for the catalog to populate.
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-device/src/device.js]: This comment is no longer valid.

File: packages/node_modules/@webex/internal-plugin-device/src/device.js [hunk 4]
```diff
@@ -468,38 +468,25 @@ const Device = WebexPlugin.extend({
...
-    ));
+    return services.waitForCatalog(
+      'postauth',
+      this.config.canRegisterWaitDuration
+    )
+      .then(() => (
+        // Validate that the service exists after waiting for the catalog.
+        services.get('wdm') ?
+          Promise.resolve() :
+          Promise.reject(new Error('`wdm` service is not available'))
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-device/src/device.js]: Can we update the error message a bit:

```suggestion
          Promise.reject(new Error('device: cannot register, `wdm` service is not available'))
```

File: packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js (hunk 1 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: implemented.

File: packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js (hunk 2 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: implemented.

File: packages/node_modules/@webex/internal-plugin-device/src/device.js (hunk 3 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-device/src/device.js]: implemented.

File: packages/node_modules/@webex/internal-plugin-device/src/device.js (hunk 4 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-device/src/device.js]: implemented.

File: packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js [hunk 5]
```diff
@@ -67,135 +67,64 @@ describe('plugin-device', () => {
...
-              services.waitForCatalog = sinon.stub().resolves();
-            });
-
-            it('should return a rejected promise', () => {
-              /* eslint-disable-next-line no-underscore-dangle */
-              assert.isTrue(services._getCatalog().status.postauth.collecting);
-              assert.isRejected(device.canRegister(wait));
+          return services.waitForCatalog('postauth')
+            .then(() => {
+              assert.isDefined(services.get('wdm'));
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: Can you remove this assert that lives in the `beforeEach` block?

Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-device/test/integration/spec/device.js]: implemented.

**Pull request merged**